support-workers/src/main/scala/com/gu/zuora/ProductSubscriptionBuilders.scala

```diff
@@ -68,53 +74,74 @@ object ProductSubscriptionBuilders {
         contractEffectiveDate = contractEffectiveDate
       )
 
-      val redeemedSub = paymentMethod.fold(
-        _ => subscriptionData.subscription,
-        redemptionData => redemptionData.redeem(subscriptionData.subscription)
+      paymentMethod.fold({ case (_, maybePromoCode) =>
+        EitherT.fromEither[Future](
+          applyPromoCode(promotionService, maybePromoCode, country, productRatePlanId, subscriptionData)
+            .left.map(_.msg)
+        )
+        },
+        redemptionData =>
+          withRedemption(subscriptionData.subscription, redemptionData, getCodeStatus)
+            .leftMap(_.clientCode).map(subscription => subscriptionData.copy(subscription = subscription))
       )
 
-      applyPromoCode(promotionService, maybePromoCode, country, productRatePlanId, subscriptionData.copy(subscription = redeemedSub))
     }
 
-    def maybeRedeemCode(maybeRedemptionData: Option[RedemptionData], subscriptionData: SubscriptionData): SubscriptionData = {
-      val redeemedSub = maybeRedemptionData.map(_.redeem(subscriptionData.subscription)).getOrElse(subscriptionData.subscription)
-      subscriptionData.copy(subscription = redeemedSub)
+    def withRedemption(
+      subscription: Subscription,
+      redemptionData: RedemptionData,
+      getCodeStatus: GetCodeStatus
+    )(implicit ec: ExecutionContext): EitherT[Future, GetCodeStatus.RedemptionInvalid, Subscription] = {
+      val withCode = subscription.copy(redemptionCode = Some(redemptionData.redemptionCode))
+      redemptionData match {
+        case CorporateRedemption(redemptionCode, _) =>
```

Comment:
here we ignore the corporate id in the payload in favour of the one that comes back from getCodeStatus.  I am not sure if the one in the payload is trusted or whether it came from the client side?

Reply:
I believe it comes from the client so best to use the one from getCodeStatus